fix(ingest): improve performance of get_allowed_list in AllowDenyPattern when dealing with large lists

Masterchen09 opened the pull request:
When dealing with large allow pattern lists get_allowed_list is extremely inefficient, because each allow pattern is validated against all other allow pattern - as we are already dealing with all allowed patterns in the list we don't have to validate them whether they are allowed, because they are...we only have to check whether they are denied or not.

Example: For a list of 2730 allow patterns up to 2730^2 resp. 7.452.900 unnecessary regular expression checks are performed.

This PR adds a new method "denied" to AllowDenyPattern which is used to filter the allow patterns in get_allowed_list.

Of course this logic only works in case the allow list is a fully specified allow list, but this is already checked/asserted* in get_allowed_list. It seems currently the get_allowed_list is not used anywhere by the acryl-datahub package, but for a custom source we have large list of allow patterns, which can currently not be used effectively because of these unneccessary checks.

*Side-note: Should the assert statement really be used in productive code? I don't think so because in case optimization is used the assert statements are ignored and strange things might happen then...but there are also other locations where the assert statement is used, therefore that's maybe something which should be addressed in general and not within this PR...

See also the Python docs regarding the assert statement:
https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement
https://docs.python.org/3/using/cmdline.html#cmdoption-O

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

hsheth2 commented on metadata-ingestion/src/datahub/configuration/common.py:
```diff
@@ -243,15 +243,21 @@ def allow_all(cls) -> "AllowDenyPattern":
             for allow_pattern in self.allow
         )
 
+    def denied(self, string: str) -> bool:
```
we probably don't want this to be part of the "public" interface

```suggestion
    def _denied(self, string: str) -> bool:
```

hsheth2 reviewed (commented):
Re asserts - you're right that this case should probably be changed to a proper if statement + raise

That said, our biggest usages of asserts are in test code (where pytest does the right thing) and with isinstance, where it's mainly a hint for mypy and is ok if it doesn't run in with optimization enabled

Masterchen09 commented:
You're right, I thought as the allowed method is public, then the denied method could also be public, but when do you really want to check if something is only denied and not allowed - I have changed it. :-)

Regarding the assertion this is also changed in this case - not sure, but I think a ValueError fits - however there are not only assertions in tests...especially some of the ingestion sources are using assertions a lot: https://github.com/search?q=repo%3Adatahub-project%2Fdatahub+"assert+"+language%3APython+path%3A%2F^metadata-ingestion\%2Fsrc\%2F%2F&type=code

hsheth2 reviewed (approved):
On the `assert` thing - you're totally right, and those should probably be replaced with ValueErrors or similar. Even more importantly, it looks like there's a few places where we use `AssertionError`s for control flow which definitely feels error-prone

That said, I don't know of anyone using python's `-O` with datahub, so it hasn't come up before
